haystack/agents/base.py

```diff
@@ -129,60 +139,39 @@ def run(
                         `{"Retriever": {"top_k": 10}, "Reader": {"top_k": 3}}`.
                         Parameters can only be passed to tools that are pipelines but not nodes.
         """
-        if len(self.tools) == 0:
+        if not self.tools:
             raise AgentError(
                 "Agents without tools cannot be run. Add at least one tool using `add_tool()` or set the parameter `tools` when initializing the Agent."
             )
         if max_iterations is None:
             max_iterations = self.max_iterations
 
-        transcript = ""
-        # Generate string representations of the tools to be passed to the PromptTemplate of the Agent
-        tool_names = ", ".join(self.tools.keys())
-        tool_names_with_descriptions = "\n".join([f"{tool.name}: {tool.description}" for tool in self.tools.values()])
-
+        transcript = self._get_initial_transcript(query=query)
         # Generate a thought with a plan what to do, choose a tool, generate input for it, and run it.
         # Repeat this until the final answer is found or the maximum number of iterations is reached.
         for _ in range(max_iterations):
-            pred = self.prompt_node.prompt(
-                prompt_template=self.prompt_template,
-                query=query,
-                tool_names=tool_names,
-                tool_names_with_descriptions=tool_names_with_descriptions,
-                generated_text=transcript,
-            )[0]
-
-            # Try to extract the final answer or a tool name and tool input from the generated text
+            pred = self.prompt_node(transcript)[0]
```

Comment:
This will fail if an empty list is returned, am not sure how likely that is but I would raise a clear exception.